alibi/prototypes/protoselect.py

```diff
@@ -0,0 +1,543 @@
+                 lbd: Optional[float] = None,
+                 batch_size: int = int(1e10),
+                 preprocess_fn: Optional[Callable[[np.ndarray], np.ndarray]] = None,
+                 verbose: bool = False,
+                 **kwargs):
+        """
+        Constructor
+
+        Parameters
+        ----------
+        kernel_distance
+            Kernel distance to be used. Expected to support computation in batches.
+        eps
+            Epsilon ball size.
+        lbd
+            Penalty for each prototype. Encourages a lower number of prototypes to be selected.
+        batch_size
+            Batch size to be used for kernel matrix computation.
+        preprocess_fn
+            Preprocessing function for kernel matrix computation.
+        verbose
+             Whether to display progression bar while computing prototypes points.
+        """
+        super().__init__(meta=deepcopy(DEFAULT_META_PROTOSELECT))
+        self.kernel_distance = kernel_distance
+        self.eps = eps
+        self.lbd = lbd
+        self.batch_size = batch_size
+        self.preprocess_fn = preprocess_fn
+        self.verbose = verbose
+
+        # get kernel tag
+        if hasattr(self.kernel_distance, '__name__'):
+            kernel_distance_tag = self.kernel_distance.__name__
+        elif hasattr(self.kernel_distance, '__class__'):
+            kernel_distance_tag = self.kernel_distance.__class__.__name__
+        else:
+            kernel_distance_tag = 'unknown kernel distance'
+
+        # update metadata
+        self.meta['params'].update({
+            'kernel_distance': kernel_distance_tag,
+            'eps': eps,
+            'lbd': lbd,
+            'batch_size': batch_size,
+            'verbose': verbose
+        })
+
+    def fit(self,
+            X: np.ndarray,
+            X_labels: Optional[np.ndarray] = None,
+            Y: Optional[np.ndarray] = None) -> 'ProtoSelect':
+        """
+        Fit the explainer by setting the reference dataset.
+
+        Parameters
+        ---------
+        X
+            Reference dataset to be summarized.
+        X_labels
+            Labels of the reference dataset.
+        Y
+            Dataset to choose the prototypes from. If ``None``, the prototypes will be selected from the reference
+            dataset `X`.
+
+        Returns
+        -------
+        self
+            Reference to itself.
+        """
+        self.X = X
+        # if the `X_labels` are not provided, then consider that all elements belong to the same class. This means
+        # that loss term which tries to avoid including in an epsilon ball elements belonging to other classes
+        # will always be 0. Still the first term of the loss tries to cover as many examples as possible with
+        # minimal overlap between the epsilon balls corresponding to the other prototypes.
+        self.X_labels = X_labels.astype(np.int32) if (X_labels is not None) else np.zeros((len(X), ), dtype=np.int32)
```

Comment:
Perhaps we should validate first that the shape/length of `X_labels` is what we expect, given the shape/length of `X`.

Reply:
Fixed and included a test.